test/distributed/optim/test_zero_redundancy_optimizer.py

```diff
@@ -224,24 +228,46 @@ def test_step(self):
             m.bias.data = torch.tensor([2.0])
             m.to(self.device)
 
-            o = ZeroRedundancyOptimizer(m.parameters(), optimizer_class=SGD, lr=0.1)
+            lr = 0.1
+            o = ZeroRedundancyOptimizer(m.parameters(), optimizer_class=SGD, lr=lr)
             y = m(x)
             y.backward(x)
             for p in m.parameters():
                 dist.all_reduce(p.grad.data, op=dist.ReduceOp.SUM)
                 p.grad.data /= self.world_size
             o.step()
-            self.assertEqual(m.weight, torch.tensor([[0.75]], device=self.device))
-            self.assertEqual(m.bias, torch.tensor([1.85], device=self.device))
+
+            # for y = wx + b, dy / dw = x and dy / db = 1
+            # however, passing x into y.backward() means an extra factor of x
+            # let k = world_size; then, x_i = i + 1 for i in {0, ..., k - 1}
+            # weight update:
+            #     w = w - lr * (sum_{i=1}^{k} x_i^2) / k
+            #       = w - lr * k(k + 1)(2k + 1) / (6k)
+            #       = w - lr * (k + 1)(2k + 1) / 6
+            # bias update:
+            #     b = b - lr * (sum_{i=1}^{k} x_i) / k
+            #       = b - lr * k(k + 1) / (2k)
+            #       = b - lr * (k + 1) / 2
+            self.assertEqual(
+                m.weight,
+                torch.tensor([[1.0 - lr * (self.world_size + 1) *
+                              (2 * self.world_size + 1) / 6]],
+                             device=self.device)
+            )
+            self.assertEqual(
+                m.bias,
+                torch.tensor([2.0 - lr * (self.world_size + 1) / 2],
+                             device=self.device))
+
 
     @common_distributed.skip_if_rocm
     def test_step_with_closure(self):
```

Comment:
The gradient updates for this test are independent on the world size given how the closure is defined. Hence, no generalization is needed like it was for `test_step()`.